lib/blocks.js

```diff
@@ -7,242 +7,207 @@ const _ = require('lodash');
+  // just leaking the results from an old `async.auto`?
+  const eventHash = records.map(e => e.meta.eventHash);
+  const cache = await _updateCache({hashes: eventHash, ledgerNodeId});
+  return {eventHash, cache};
+});
 
 // TODO: document
 // consensusResult = {event: [event records], consensusProof: [event records]}
-api.write = ({ledgerNode, state, consensusResult}, callback) => {
+api.write = callbackify(async ({ledgerNode, state, consensusResult}) => {
   const {blockHeight, previousBlockId: previousBlock, previousBlockHash} =
     state;
   const ledgerNodeId = ledgerNode.id;
-  async.auto({
-    config: callback =>
-      ledgerNode.storage.events.getLatestConfig((err, result) => {
-        if(err) {
-          return callback(err);
-        }
-        const config = result.event.ledgerConfiguration;
-        if(config.consensusMethod !== 'Continuity2017') {
-          return callback(new BedrockError(
-            'Consensus method must be "Continuity2017".',
-            'InvalidStateError', {
-              consensusMethod: config.consensusMethod
-            }));
-        }
-        callback(null, config);
-      }),
-    keys: ['config', (results, callback) => {
-      const {ledger: ledgerId} = results.config;
-      const blockId = _generateBlockId({blockHeight, ledgerId});
-      const signatureCreators = consensusResult.creators;
-      async.map(
-        signatureCreators, (publicKeyId, callback) =>
-          _storage.keys.getPublicKey({ledgerNodeId, publicKeyId}, callback),
-        (err, result) => {
-          if(err) {
-            return callback(err);
+  const {event: {ledgerConfiguration: ledgerConfig}} =
+    await ledgerNode.storage.events.getLatestConfig();
+  if(ledgerConfig.consensusMethod !== 'Continuity2017') {
+    throw new BedrockError(
+      'Consensus method must be "Continuity2017".',
+      'InvalidStateError', {
+        consensusMethod: ledgerConfig.consensusMethod
+      });
+  }
+
+  // FIXME: remove keys from blocks; unnecessary
+  const {ledger: ledgerId} = ledgerConfig;
+  const blockId = _generateBlockId({blockHeight, ledgerId});
+  const signatureCreators = consensusResult.creators;
+  const keys = await Promise.all(
+    signatureCreators.map(publicKeyId =>
+      _storage.keys.getPublicKey({ledgerNodeId, publicKeyId})));
+  // `seeAlso` may have been updated during a prior failed attempt to
+  // write the block
+  const publicKey = keys.map(key =>
+    (key.seeAlso && key.seeAlso !== blockId) ?
+      {id: key.id, seeAlso: key.seeAlso} : {
+        id: key.id,
+        type: key.type,
+        owner: key.owner,
+        publicKeyBase58: key.publicKeyBase58
+      });
+
+  // update events with block information 100 at a time
+  const hashes = consensusResult.eventHash;
+  let now = Date.now();
+  const limit = pLimit(100);
+  await Promise.all(hashes.map((eventHash, index) => {
+    limit(() => ledgerNode.storage.events.update({
```

Comment:
Looks like it should return the resulting promise from the limit(...) call, or just remove the surrounding `{...}`.